pkg/expr/nodes.go

```diff
@@ -229,70 +229,89 @@ func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s
+	}
 
-			// Handle Numeric Table
-			if frame.TimeSeriesSchema().Type == data.TimeSeriesTypeNot && isNumberTable(frame) {
-				logger.Debug("expression datasource query (numberSet)", "query", refID)
-				numberSet, err := extractNumberSet(frame)
-				if err != nil {
-					return mathexp.Results{}, err
-				}
-				for _, n := range numberSet {
-					vals = append(vals, n)
-				}
+	dataSource := dn.datasource.Type
+	if isAllFrameVectors(dataSource, response.Frames) { // Prometheus Specific Handling
+		vals, err = framesToNumbers(response.Frames)
+		if err != nil {
+			return mathexp.Results{}, fmt.Errorf("failed to read frames as numbers: %w", err)
+		}
+		responseType = "vector"
+		return mathexp.Results{Values: vals}, nil
+	}
 
-				return mathexp.Results{
-					Values: vals,
-				}, nil
-			}
+	if len(response.Frames) == 1 {
+		frame := response.Frames[0]
+		// Handle Untyped NoData
+		if len(frame.Fields) == 0 {
+			return mathexp.Results{Values: mathexp.Values{mathexp.NoData{Frame: frame}}}, nil
 		}
 
-		for _, frame := range qr.Frames {
-			logger.Debug("expression datasource query (seriesSet)", "query", refID)
-			// Check for TimeSeriesTypeNot in InfluxDB queries. A data frame of this type will cause
-			// the WideToMany() function to error out, which results in unhealthy alerts.
-			// This check should be removed once inconsistencies in data source responses are solved.
-			if frame.TimeSeriesSchema().Type == data.TimeSeriesTypeNot && dataSource == datasources.DS_INFLUXDB {
-				logger.Warn("ignoring InfluxDB data frame due to missing numeric fields", "frame", frame)
-				continue
-			}
-			series, err := WideToMany(frame)
+		// Handle Numeric Table
+		if frame.TimeSeriesSchema().Type == data.TimeSeriesTypeNot && isNumberTable(frame) {
+			numberSet, err := extractNumberSet(frame)
 			if err != nil {
 				return mathexp.Results{}, err
 			}
-			for _, s := range series {
-				vals = append(vals, s)
+			for _, n := range numberSet {
+				vals = append(vals, n)
 			}
+			responseType = "number set"
+			return mathexp.Results{
+				Values: vals,
+			}, nil
 		}
 	}
+
+	for _, frame := range response.Frames {
+		// Check for TimeSeriesTypeNot in InfluxDB queries. A data frame of this type will cause
+		// the WideToMany() function to error out, which results in unhealthy alerts.
+		// This check should be removed once inconsistencies in data source responses are solved.
+		if frame.TimeSeriesSchema().Type == data.TimeSeriesTypeNot && dataSource == datasources.DS_INFLUXDB {
+			logger.Warn("Ignoring InfluxDB data frame due to missing numeric fields")
+			continue
+		}
+		series, err := WideToMany(frame)
+		if err != nil {
+			return mathexp.Results{}, err
+		}
+		for _, s := range series {
+			vals = append(vals, s)
+		}
+	}
+
+	responseType = "series set"
 	return mathexp.Results{
-		Values: vals,
+		Values: vals, // TODO vals can be empty. Should we replace with no-data?
```

Comment:
@kylebrandt, @grobinson-grafana  wdyt?

Reply:
Not sure, probably best to keep it as now. I'm okay with leaving the TODO, I am going to be looking at adopting SSE to the dataplane contract in the coming weeks, so will be around this code again shortly.